### PR TITLE
add channels last support for AdaptiveAvgPool2d on CPU path

### PR DESCRIPTION
**Pull request by mingfeima**
This patch adds channels last support for `nn.AdaptiveAvgPool2d`, both forward and backward included.

### CODE REVIEW

**Comment by mingfeima** [comment]
#### Updates:
1. Add support for `ChannelsLast` memory format on CPU, this path is vectorized upon the dimension of **channels**.
2. Move `Contiguous` path to native/cpu.

`adaptive_avg_pool2d` has a fast path when output_size is 1x1 for contiguous memory format, this patch did not change that. Similar approach on channels last requires a `reshape` which makes it less performant, since the generic kernel `adaptive_avg_pool2d_kernel` on channels last already performs better than the fast path, i skip to the implement fast path for channels last when output size is 1x1.

#### Results:
**Machine**: CPU Intel(R) Xeon(R) Platinum 8260 CPU @ 2.40GHz, 2*20 cores.
**Bench**: Use this [script](https://github.com/mingfeima/op_bench-py/blob/master/adaptive_avg_pool2d.py) to reproduce, `./run.sh adaptive_avg_pool2d.py`.
**Input size**: [1, 2048, 7, 7], [128, 2048, 7, 7]: (size in ResNet50).
**Output size**: [1, 1], [2, 2]
Both single thread (1 core) and single socket (20 core) are tested.

Code base: before: [96aaa311](https://github.com/mingfeima/pytorch/commit/96aaa311c0251d24decb9dc5da4957b7c590af6f), after: [86bf0cc7](https://github.com/mingfeima/pytorch/commit/86bf0cc71d725b61a207033d6ab5e503e4e462c1).

#cores | input_size | output_size | before (contiguous) | after (contiguous) | after (channels_last) | cl/contig
-- | -- | -- | -- | -- | -- | --
20 | [1, 2028, 7, 7] | [2, 2] | 0.025 | 0.021 | 0.02 | 1.05
20 | [128, 2028, 7, 7] | [2, 2] | 1.973 | 1.021 | 0.373 | 2.74
20 | [1, 2028, 7, 7] | [1, 1] | 0.033 | 0.032 | 0.022 | 1.45
20 | [128, 2028, 7, 7] | [1, 1] | 0.44 | 0.451 | 0.311 | 1.45
1 | [1, 2028, 7, 7] | [2, 2] | 0.22 | 0.146 | 0.034 | 4.29
1 | [128, 2028, 7, 7] | [2, 2] | 26.806 | 16.949 | 5.309 | 3.19
1 | [1, 2028, 7, 7] | [1, 1] | 0.037 | 0.037 | 0.016 | 2.31
1 | [128, 2028, 7, 7] | [1, 1] | 4.649 | 4.605 | 3.626 | 1.27

**Comment by mingfeima** [comment]
@ngimel  could you please enlighten me how to enable the nhwc test for CPU at `test_nn.py`, for example this one below:
test_adaptive_pooling_avg_nhwc: https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L3472

**Comment by mingfeima** [comment]
@VitalyFedyunin @ngimel is it OK to use `#pragma omp simd` in ATen when trying to vectorize the scalar code?

File: aten/src/ATen/native/AdaptiveAveragePooling.cpp [hunk 1]
```diff
@@ -1,303 +1,66 @@
...
-      }
-    });
-  }
-
   void adaptive_avg_pool2d_out_cpu_template(
     at::Tensor& output,
     at::Tensor const& input,
     IntArrayRef output_size)
   {
-    TORCH_CHECK(output_size.size() == 2, "adaptive_avg_pool2d: output_size must be 2");
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/AdaptiveAveragePooling.cpp:109]: Please do not drop this check

File: aten/src/ATen/native/AdaptiveAveragePooling.cpp [hunk 2]
```diff
@@ -1,303 +1,66 @@
...
-    for (int64_t i = 0; i < input.ndimension(); i++) {
+    int64_t ndim = input.ndimension();
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/AdaptiveAveragePooling.cpp:16]: Please check if dtype of the output and dtype of the input are the same

File: aten/src/ATen/native/AdaptiveAveragePooling.cpp [hunk 3]
```diff
@@ -1,303 +1,66 @@
...
-    /* sizes */
-    int sizeD  = input.size(-3);
-    int isizeH = input.size(-2);
-    int isizeW = input.size(-1);
-    int osizeH = gradOutput_.size(-2);
-    int osizeW = gradOutput_.size(-1);
-
-    /* get contiguous gradOutput */
-    auto gradOutput = gradOutput_.contiguous();
+    int64_t ndim = grad_output.ndimension();
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/AdaptiveAveragePooling.cpp:48]: Please check if all args dtypes are the same

File: aten/src/ATen/native/cpu/AdaptiveAvgPoolKernel.cpp [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+      for (; d3 < size; d3++) {
+        out[d3] = out[d3] / kh / kw;
+      }
+
+      // move on to next output index
+      data_index_step(n, nbatch, oh, output_height, ow, output_width);
+    }
+  });
+
+  if (!output_.is_contiguous(memory_format)) {
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cpu/AdaptiveAvgPoolKernel.cpp:147]: Thank you for supporting strided outputs, but could you also add test for this case.

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Overall looks good (I still need to dive into correctness). Please add tests for various memory format input and outputs.

File: aten/src/ATen/native/AdaptiveAveragePooling.cpp (hunk 1 above)
Comment (mingfeima) [reply aten/src/ATen/native/AdaptiveAveragePooling.cpp:109]: fixed

File: aten/src/ATen/native/AdaptiveAveragePooling.cpp (hunk 2 above)
Comment (mingfeima) [reply aten/src/ATen/native/AdaptiveAveragePooling.cpp:17]: fixed

File: aten/src/ATen/native/AdaptiveAveragePooling.cpp (hunk 3 above)
Comment (mingfeima) [reply aten/src/ATen/native/AdaptiveAveragePooling.cpp:51]: fixed

File: aten/src/ATen/native/cpu/AdaptiveAvgPoolKernel.cpp (hunk 4 above)
Comment (mingfeima) [reply aten/src/ATen/native/cpu/AdaptiveAvgPoolKernel.cpp:147]: `test_adaptive_pooling_avg_nhwc` and `test_adaptive_pooling_avg_nhwc_non_contiguous` cover both device `cpu` and `gpu` now.

**Comment by mingfeima** [comment]
@VitalyFedyunin test cases added for cpu path on channels last memory format, please review. Sorry for the delay, got distracted by some other job.
I will rebase other relevant PRs upon this one, #42719, #46234, etc.

**Comment by mingfeima** [comment]
please review the new one #48916

**Pull request closed without merging**